### PR TITLE
feat: duplicate products

### LINKED ISSUE #8157: Duplicate product
### Description

This task is about adding a "duplicate product" feature, already available on iOS:

https://github.com/woocommerce/woocommerce-ios/issues/7255

### PR DESCRIPTION
**Pull request by wzieba**
Closes: #8157

### Description

This PR adds a feature of duplicating products - including variable products.

iOS reference: https://github.com/woocommerce/woocommerce-ios/pull/7727

Firstly review&merge FluxC PR: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2629

### Testing instructions

The feature is self-explanatory, please test it as you were a user. Interesting corner cases:
- duplicate product that has more than 100 variations
- verify that if you customize a variation (add description or change image) and then you duplicate a product, variations in duplicated product are also customized (have the same description or image). In other words, please check if https://github.com/woocommerce/woocommerce-ios/issues/8611 doesn't happen.

### Images/gif

https://user-images.githubusercontent.com/5845095/212022201-dcea928f-b066-4d63-b162-29cf08421927.mp4

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/DuplicateProduct.kt [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    private val variationRepository: VariationRepository,
+    private val resourceProvider: ResourceProvider,
+) {
+
+    suspend operator fun invoke(product: Product): Result<Long> {
+        val newProduct = product.copy(
+            remoteId = 0,
+            name = resourceProvider.getString(R.string.product_duplicate_copied_product_name, product.name),
+            sku = "",
+            status = ProductStatus.DRAFT
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/DuplicateProduct.kt:24]: This follows the same principles as iOS implementation. 
- remoteId has to be 0 no Core will create a product
- new name localized
- SKU is a unique property.
- status after duplication is draft

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/DuplicateProduct.kt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        )
+
+        val (duplicateProductSuccess, duplicatedProductRemoteId) = productDetailRepository.addProduct(newProduct)
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/DuplicateProduct.kt:27]: 🌱 I do not like the API of `ProductDetailRepository#addProduct`, maybe we could improve on the next occasion to use `kotlin.Result`?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/DuplicateProduct.kt [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        }
+    }
+
+    private suspend fun duplicateVariations(
+        product: Product,
+        duplicatedProductRemoteId: Long
+    ): Result<Long> {
+        val duplicatedVariations = variationRepository.getAllVariations(product.remoteId).map {
+            it.copy(remoteProductId = duplicatedProductRemoteId, sku = "")
+        }
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/DuplicateProduct.kt:54]: ⚠️ I don't trust `VariationRepository#getAllVariations` as it uses the state of a repository (`offset` and `canLoadMoreProductVariations`).

While I was testing the feature, once or twice not all variations were duplicated. 25 were missing (25 is page size when fetching variations). But I can't reproduce it.

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/products/DuplicateProductTest.kt [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        val duplicationRequestCapture = argumentCaptor<Product>()
+        verify(productDetailRepository).addProduct(duplicationRequestCapture.capture())
+
+        assertThat(duplicationRequestCapture.firstValue)
+            .matches {
+                it.remoteId == 0L && it.name == "copied name" && it.sku == "" && it.status == ProductStatus.DRAFT
+            }
+            .usingRecursiveComparison()
+            .ignoringFields("remoteId", "name", "sku", "status")
+            .isEqualTo(productToDuplicate)
```
Comment (atorresveiga) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/ui/products/DuplicateProductTest.kt:61]: Nice!

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/DuplicateProduct.kt (hunk 3 above)
Comment (atorresveiga) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/DuplicateProduct.kt:54]: > While I was testing the feature, once or twice, not all variations were duplicated.

I also tried but could not reproduce this, so I think this issue is not a blocker for merging the feature.
That said, I agree with you that we can refactor getAllVariations with a more reliable form for getting all the data. Something that is out of the scope of this PR.

**Review by atorresveiga — APPROVED** [review]
Great work @wzieba! 
I tested the code and everything is working as expected, even variable products with more than 100 variations (I tested with 101 variations).
LGTM :shipit: 
The PR still has the `do not merge` tag. Is there anything else you would like to add to this PR?

**Comment by wzieba** [comment]
Thank you for the review @atorresveiga !

> Is there anything else you would like to add to this PR?

I'll update FluxC hash and merge this PR 🙂

**Comment by wzieba** [comment]
FYI @atorresveiga - I'll wait a little longer with merge for release 12.0 to start its lifecycle
